File: cellab_devkit-1.0 (1)/ruledefs/timetun.js
/*

        Second-order dynamics: The time tunnel

        (Toffoli & Margolus, section 6.3, page 52)

*/

    rule.worldtype = 1;          // 2D torus world
    rule.patreq = "timetun";
    rule.palreq = "default";


    function timetun(oldstate,     nw, n  , ne,
                                   w, self, e,
                                   sw, s  , se) {
        var ttmap = [ 0, 1, 1, 1, 1, 0 ];
        
        return (ttmap[self + n + s + w + e] ^ ((oldstate >> 1) & 1)) |
                (self << 1);
    }

File: cellab_devkit-1.0 (1)/rules_java/timetun.java
/*

        Second-order dynamics: The time tunnel

        (Toffoli & Margolus, section 6.3, page 52)

*/

class Timetun extends ruletable {
    static final int ttmap[] = {0, 1, 1, 1, 1, 0};

    void jcruleModes() {
        setWorld(1);
        setPatternRequest("timetun");
        setPaletteRequest("default");
    }

    int jcrule(int oldstate) {
        return (ttmap[self + n + s + w + e] ^ ((oldstate >> 1) & 1)) |
                (self << 1);
    }
}

public class timetun {
    public static void main(String args[]) {
        (new Timetun()).generateRuleFile("timetun");
    }
}
